core/src/container.cpp

```diff
@@ -487,7 +487,8 @@ struct SolutionCollector
 			solutions.emplace_back(std::make_pair(trace, prio));
 		} else {
 			for (SolutionBase* successor : next) {
-				assert(!successor->isFailure());  // We shouldn't have invalid solutions
+				if (successor->isFailure())
+					continue;  // skip failures
```

Comment:
I tried to get my head around why this happens in some situations I encountered in the past and failed.
Can you describe a scenario where the assert will be violated with pruning active?
I'm fine with changing it because it is a step necessary if we want to have a flag that disables pruning for a whole task (any unpruned stage with a dead end on the other side of a partial solution will trigger this assert), but I do not yet understand why it can be invalid.

Reply:
I'm not sure about the reasons anymore (it's nearly half a year ago that I handled this). 
I think we cannot be sure that only valid solutions enter the start/end interfaces. For example, Connect stages do preserve all InterfaceStates, both successes and failures. We just save failures for later inspection as well.
Thus, we need to explicitly filter them here.
I remember that this was a no-brainer when encountering it. But now I don't remember a concrete example.

Reply:
> it's nearly half a year ago that I handled this

Fair enough, my fault for letting it lie around for so long.

> For example, Connect stages do preserve all InterfaceStates, both successes and failures.

Of course, but they also *prune* the paths on failure. so unless the solution path forks, the other end should not compute at all (as the interface state is pruned). I came up with an example case below where the assert is violated with forking solution paths:
![g3811](https://github.com/ros-planning/moveit_task_constructor/assets/680358/8e95536a-7605-4dea-b120-00f63198634c)

But I observed a rare stochastic assert in the following situation some days ago where `turn` yielded a solution which hit the assert when propagating back the priority to the failed attempt in `approach grasp`:
![screenshot240214-173210-red](https://github.com/ros-planning/moveit_task_constructor/assets/680358/183f7da6-bac5-4d49-b4fe-4603d47daa86)
In this situation there are no obvious forking solution paths (unless ComputeIK factors through a single pose in the wrapped generator, but that should not happen?) and the failure in `approach grasp` should have pruned the only connected interface start state in `turn`, so the compute call that generated the `turn` solution should not have happened.
I tried to write a test case for this, but failed to do so in reasonable time.

So TLDR: I guess dropping the assert is correct, but also obfuscates bugs in the pruning.

Reply:
Strange that you cannot reproduce the error in a simpler unit test.